File: app/local_plugins/core/local_db/ios/Classes/LocalDbPlugin.swift
import Flutter

public class LocalDbPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private let db = AppDatabase.shared

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = LocalDbPlugin()
        instance.channel = FlutterMethodChannel(
            name: "local_db/commands",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: instance.channel!)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            do {
                let value: Any? = try self.dispatch(call.method, args: args)
                DispatchQueue.main.async { result(value) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DB_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func dispatch(_ method: String, args: [String: Any]) throws -> Any? {
        switch method {

        // ── ServiceConfig ────────────────────────────────────────────────
        case "upsertServiceConfig":
            try db.upsertServiceConfig(ServiceConfigRecord(
                id: args["id"] as! String,
                type: args["type"] as! String,
                vendor: args["vendor"] as! String,
                name: args["name"] as! String,
                configJson: args["configJson"] as! String,
                createdAt: Int64(args["createdAt"] as! Int)
            ))
            return nil

        case "deleteServiceConfig":
            try db.deleteServiceConfig(id: args["id"] as! String)
            return nil

        case "getAllServiceConfigs":
            return try db.getAllServiceConfigs().map { e in
                [
                    "id": e.id, "type": e.type, "vendor": e.vendor,
                    "name": e.name, "configJson": e.configJson,
                    "createdAt": e.createdAt,
                ] as [String: Any]
            }

        // ── Agent ────────────────────────────────────────────────────────
        case "upsertAgent":
            try db.upsertAgent(AgentRecord(
                id: args["id"] as! String,
                name: args["name"] as! String,
                type: args["type"] as! String,
                configJson: args["configJson"] as! String,
                createdAt: Int64(args["createdAt"] as! Int),
                updatedAt: Int64(args["updatedAt"] as! Int)
            ))
            return nil

        case "deleteAgent":
            try db.deleteAgent(id: args["id"] as! String)
            return nil

        case "getAllAgents":
            return try db.getAllAgents().map { e in
                [
                    "id": e.id, "name": e.name, "type": e.type,
                    "configJson": e.configJson,
                    "createdAt": e.createdAt, "updatedAt": e.updatedAt,
                ] as [String: Any]
            }

        // ── Message ──────────────────────────────────────────────────────
        case "getMessages":
            let agentId = args["agentId"] as! String
            let limit = args["limit"] as! Int
            return try db.getMessages(agentId: agentId, limit: limit).map { e in
                [
                    "id": e.id, "agentId": e.agentId, "role": e.role,
                    "content": e.content, "status": e.status,
                    "createdAt": e.createdAt, "updatedAt": e.updatedAt,
                ] as [String: Any]
            }

        case "deleteMessages":
            try db.deleteMessagesByAgent(agentId: args["agentId"] as! String)
            return nil

        // ── McpServer ────────────────────────────────────────────────────
        case "upsertMcpServer":
            try db.upsertMcpServer(McpServerRecord(
                id: args["id"] as! String,
                agentId: args["agentId"] as! String,
                name: args["name"] as! String,
                url: args["url"] as! String,
                transport: args["transport"] as! String,
                authHeader: args["authHeader"] as? String,
                enabledToolsJson: args["enabledToolsJson"] as! String,
                isEnabled: args["isEnabled"] as! Bool,
                createdAt: Int64(args["createdAt"] as! Int)
            ))
            return nil

        case "deleteMcpServer":
            try db.deleteMcpServer(id: args["id"] as! String)
            return nil

        case "getMcpServersByAgent":
            let agentId = args["agentId"] as! String
            return try db.getMcpServersByAgent(agentId: agentId).map { e in
                [
                    "id": e.id, "agentId": e.agentId, "name": e.name,
                    "url": e.url, "transport": e.transport,
                    "authHeader": e.authHeader as Any,
                    "enabledToolsJson": e.enabledToolsJson,
                    "isEnabled": e.isEnabled, "createdAt": e.createdAt,
                ] as [String: Any]
            }

        default:
            return FlutterMethodNotImplemented
        }
    }
}

File: app/local_plugins/core/local_db/android/src/main/kotlin/com/aiagent/local_db/LocalDbPlugin.kt
package com.aiagent.local_db

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*
import com.aiagent.local_db.entity.*

class LocalDbPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private lateinit var db: AppDatabase
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        db = AppDatabase.getInstance(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, "local_db/commands")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        scope.launch {
            try {
                val value = when (call.method) {
                    // ── ServiceConfig ──────────────────────────────────────
                    "upsertServiceConfig" -> {
                        val args = call.arguments as Map<*, *>
                        db.serviceConfigDao().upsert(
                            ServiceConfigEntity(
                                id = args["id"] as String,
                                type = args["type"] as String,
                                vendor = args["vendor"] as String,
                                name = args["name"] as String,
                                configJson = args["configJson"] as String,
                                createdAt = (args["createdAt"] as Number).toLong(),
                            )
                        )
                        null
                    }
                    "deleteServiceConfig" -> {
                        val args = call.arguments as Map<*, *>
                        db.serviceConfigDao().delete(args["id"] as String)
                        null
                    }
                    "getAllServiceConfigs" -> {
                        db.serviceConfigDao().getAll().map { e ->
                            mapOf(
                                "id" to e.id, "type" to e.type, "vendor" to e.vendor,
                                "name" to e.name, "configJson" to e.configJson,
                                "createdAt" to e.createdAt,
                            )
                        }
                    }

                    // ── Agent ──────────────────────────────────────────────
                    "upsertAgent" -> {
                        val args = call.arguments as Map<*, *>
                        db.agentDao().upsert(
                            AgentEntity(
                                id = args["id"] as String,
                                name = args["name"] as String,
                                type = args["type"] as String,
                                configJson = args["configJson"] as String,
                                createdAt = (args["createdAt"] as Number).toLong(),
                                updatedAt = (args["updatedAt"] as Number).toLong(),
                            )
                        )
                        null
                    }
                    "deleteAgent" -> {
                        val args = call.arguments as Map<*, *>
                        db.agentDao().delete(args["id"] as String)
                        null
                    }
                    "getAllAgents" -> {
                        db.agentDao().getAll().map { e ->
                            mapOf(
                                "id" to e.id, "name" to e.name, "type" to e.type,
                                "configJson" to e.configJson,
                                "createdAt" to e.createdAt, "updatedAt" to e.updatedAt,
                            )
                        }
                    }

                    // ── Message ────────────────────────────────────────────
                    "getMessages" -> {
                        val args = call.arguments as Map<*, *>
                        val agentId = args["agentId"] as String
                        val limit = (args["limit"] as Number).toInt()
                        db.messageDao().getMessages(agentId, limit).map { e ->
                            mapOf(
                                "id" to e.id, "agentId" to e.agentId, "role" to e.role,
                                "content" to e.content, "status" to e.status,
                                "createdAt" to e.createdAt, "updatedAt" to e.updatedAt,
                            )
                        }
                    }

                    "deleteMessages" -> {
                        val args = call.arguments as Map<*, *>
                        db.messageDao().deleteByAgent(args["agentId"] as String)
                        null
                    }

                    // ── McpServer ──────────────────────────────────────────
                    "upsertMcpServer" -> {
                        val args = call.arguments as Map<*, *>
                        db.mcpServerDao().upsert(
                            McpServerEntity(
                                id = args["id"] as String,
                                agentId = args["agentId"] as String,
                                name = args["name"] as String,
                                url = args["url"] as String,
                                transport = args["transport"] as String,
                                authHeader = args["authHeader"] as String?,
                                enabledToolsJson = args["enabledToolsJson"] as String,
                                isEnabled = args["isEnabled"] as Boolean,
                                createdAt = (args["createdAt"] as Number).toLong(),
                            )
                        )
                        null
                    }
                    "deleteMcpServer" -> {
                        val args = call.arguments as Map<*, *>
                        db.mcpServerDao().delete(args["id"] as String)
                        null
                    }
                    "getMcpServersByAgent" -> {
                        val args = call.arguments as Map<*, *>
                        db.mcpServerDao().getByAgent(args["agentId"] as String).map { e ->
                            mapOf(
                                "id" to e.id, "agentId" to e.agentId, "name" to e.name,
                                "url" to e.url, "transport" to e.transport,
                                "authHeader" to e.authHeader,
                                "enabledToolsJson" to e.enabledToolsJson,
                                "isEnabled" to e.isEnabled, "createdAt" to e.createdAt,
                            )
                        }
                    }

                    else -> {
                        withContext(Dispatchers.Main) { result.notImplemented() }
                        return@launch
                    }
                }
                withContext(Dispatchers.Main) { result.success(value) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("DB_ERROR", e.message, null)
                }
            }
        }
    }
}

// ── Data classes (kept for reference, not used by MethodChannel handler) ──

data class ServiceConfigRow(
    val id: String, val type: String, val vendor: String,
    val name: String, val configJson: String, val createdAt: Long,
)

data class AgentRow(
    val id: String, val name: String, val type: String,
    val configJson: String, val createdAt: Long, val updatedAt: Long,
)

data class MessageRow(
    val id: String, val agentId: String, val role: String,
    val content: String, val status: String,
    val createdAt: Long, val updatedAt: Long,
)

data class McpServerRow(
    val id: String, val agentId: String, val name: String,
    val url: String, val transport: String, val authHeader: String?,
    val enabledToolsJson: String, val isEnabled: Boolean, val createdAt: Long,
)
